x-pack/plugins/reporting/server/browsers/chromium/driver_factory/test.ts

```diff
@@ -0,0 +1,158 @@
+  const userDataDir = mkdtempSync(join(tmpdir(), 'chromium-'));
+  const binaryPath = getBinaryPath();
+  const kbnArgs = args({
+    userDataDir,
+    viewport: { width: 800, height: 600 },
+    disableSandbox,
+    proxy,
+  });
+  const finalArgs = uniq([...defaultArgs, ...kbnArgs, ...flags]);
+  let browserLogs = '';
+
+  const browserProcess = spawn(binaryPath, finalArgs, {
+    // On non-windows platforms, `detached: true` makes child process a
+    // leader of a new process group, making it possible to kill child
+    // process tree with `.kill(-pid)` command. @see
+    // https://nodejs.org/api/child_process.html#child_process_options_detached
+    detached: process.platform !== 'win32',
+  });
+
+  return new Promise((resolve) => {
+    const rl = createInterface({ input: browserProcess.stderr });
+
+    const timer = setTimeout(
+      () =>
+        doneOnce(
+          new Error(`Browser didn't bind successfully in ${browserLaunchTimeToWait} milliseconds`)
+        ),
+      browserLaunchTimeToWait
+    );
+
+    rl.on('line', (data) => {
+      logger.debug(`Browser log output: "${data}"`);
+      browserLogs += data + '\n';
+      const match = data.match(/^DevTools listening on/);
+      if (match) {
+        logger.info(`Browser successfully started remote DevTools listener`);
+        return doneOnce();
+      }
+    });
+
+    browserProcess.once('exit', () =>
+      doneOnce(new Error(`Browser exited abnormally during startup`))
+    );
+
+    browserProcess.on('error', () => doneOnce(new Error(`Browser errored abnormally`)));
+
+    const doneOnce = once(async (err?: Error) => {
+      if (err) {
+        logger.error(`Browser didn't start or didn't bind to the local port in time`);
+        logger.error(err);
+      }
+      const help: BrowserDiagnosticPayload['help'] = [];
+      clearTimeout(timer);
+
+      // Sometimes logs with significant issues can creep in after browser has bound it's remote
+      // debugging port, so let them flow through before SIGKILL is sent.
+      await sleep(browserCoolDown);
+
+      browserProcess.removeAllListeners();
+      rl.removeAllListeners();
+      rl.close();
+
+      for (const knownIssue of BROWSER_TEST_COMMON_ISSUES.keys()) {
+        const helpText = BROWSER_TEST_COMMON_ISSUES.get(knownIssue);
+        if (browserLogs.includes(knownIssue) && helpText) {
+          logger.warn(helpText);
+          help.push(helpText);
+        }
+      }
+
+      if (browserProcess && browserProcess.pid && !browserProcess.killed) {
+        browserProcess.kill('SIGKILL');
+        logger.info(`Successfully sent 'SIGKILL' to browser process (PID: ${browserProcess.pid})`);
+      }
+
+      logger.debug(`deleting chromium user data directory at [${userDataDir}]`);
+
+      del(userDataDir, { force: true }).catch((error) => {
+        logger.error(`error deleting user data directory at [${userDataDir}]!`);
+        logger.error(error);
+      });
+
+      return resolve({
```

Comment:
Ok if we just return the `BrowserDiagnosticPayload` object here and call `resolve` closer to where it appears first in the code? I think seeing multiple resolve calls, even though it is more verbose, helps keep in mind that there are multiple ways to resolve this promise

Reply:
I might try and mangle this into an observable since that seems more suitable/readable now that I think of it.

Reply:
++ it'll be easier to move around in different parts of browser factory / driver code if it is all the same and uses observables.